File: fight_env/orchestrator/orchestrator.py
from typing import Optional, List

from fight_env.orchestrator.duel_orchestrator import DuelOrchestrator
from fight_env.player.player import Player
from fight_env.player.refs.events import Response, Responses


class Orchestrator:
    def __init__(self, players: List[Player]):
        self.fighter1 = players[0]
        self.fighter2 = players[1]
        self.duel_orchestrator: Optional[DuelOrchestrator] = None

    def flow(self):
        snapshot1 = self.fighter1.make_snapshot()
        snapshot2 = self.fighter2.make_snapshot()

        self.fighter1.tick()
        self.fighter2.tick()

        # later distance check here
        need = True
        if need:
            self.duel_orchestrator = DuelOrchestrator(self.fighter1, self.fighter2)
            f1_responses, f2_responses = self.duel_orchestrator.resolve()
        else:
            f1_responses, f2_responses = ((Response(Responses.NONE), Response(Responses.NONE)),
                                          (Response(Responses.NONE), Response(Responses.NONE)))

        f1_fallback_resolved = self.fighter1.fallback()
        f2_fallback_resolved = self.fighter2.fallback()

        f1_intent_resolved = self.fighter1.process_intent()
        f2_intent_resolved = self.fighter2.process_intent()

        f1_reactive_resolved = self.fighter1.reactive(snapshot1)
        f2_reactive_resolved = self.fighter2.reactive(snapshot2)

        self.fighter1.cleanup(f1_fallback_resolved, f1_intent_resolved, f1_reactive_resolved)
        self.fighter2.cleanup(f2_fallback_resolved, f2_intent_resolved, f2_reactive_resolved)

        return f1_responses, f2_responses

File: unity/Assets/FightCore/Orchestrator.cs
// Port of fight_env/orchestrator/{orchestrator,duel_orchestrator}.py.
using System.Collections.Generic;

namespace FightCore
{
    public sealed class Orchestrator
    {
        private readonly Player _f1;
        private readonly Player _f2;

        public Orchestrator(Player fighter1, Player fighter2)
        {
            _f1 = fighter1;
            _f2 = fighter2;
        }

        // orchestrator.py :: Orchestrator.flow
        public (List<Response> f1Responses, List<Response> f2Responses) Flow()
        {
            var snapshot1 = _f1.MakeSnapshot();
            var snapshot2 = _f2.MakeSnapshot();

            _f1.Tick();
            _f2.Tick();

            // "later distance check here" -> currently always resolved
            var (f1Responses, f2Responses) = ResolveDuel(_f1, _f2);

            bool f1Fallback = _f1.Fallback();
            bool f2Fallback = _f2.Fallback();

            bool f1Intent = _f1.ProcessIntent();
            bool f2Intent = _f2.ProcessIntent();

            bool f1Reactive = _f1.Reactive(snapshot1);
            bool f2Reactive = _f2.Reactive(snapshot2);

            _f1.Cleanup(f1Fallback, f1Intent, f1Reactive);
            _f2.Cleanup(f2Fallback, f2Intent, f2Reactive);

            return (f1Responses, f2Responses);
        }

        // duel_orchestrator.py :: DuelOrchestrator.resolve
        private static (List<Response>, List<Response>) ResolveDuel(Player f1, Player f2)
        {
            var event1 = f1.Event();
            var event2 = f2.Event();

            var (f1Res, f2Res) = Resolution.ResolveDuelists(event1, event2);
            var (f2Res2, f1Res2) = Resolution.ResolveDuelists(event2, event1);

            var f1Responses = new List<Response> { f1Res, f1Res2 };
            var f2Responses = new List<Response> { f2Res, f2Res2 };

            f1.ProcessResponses(f1Responses);
            f2.ProcessResponses(f2Responses);

            return (f1Responses, f2Responses);
        }
    }
}